File: python/gemini/p31_nre.py
from urllib.parse import urlparse
from flask import Flask, abort, redirect, request

app = Flask(__name__)

@app.route('/redirect')
def redirect_url():
    url = request.args.get('url')

    # 1. Ensure the URL input is present
    if not url:
        return abort(404)

    try:
        # 2. The urlparse function from an imported module is used to parse the URL.
        parsed_url = urlparse(url)
        
        # 3. The network location attribute of the parsed URL object is used to get the domain name.
        domain = parsed_url.netloc

        # If the domain name is example.com the URL is redirected.
        # Strict equality check for security
        if domain == 'example.com':
            return redirect(url)
        else:
            # If the domain name is not example.com a 404 error is returned.
            return abort(404)
    except ValueError:
        # Handle parsing errors gracefully
        return abort(404)

File: javascript/gemini/p31_nre.js
const { URL } = require('url');

function redirect_url(urlInput, res) {
    // 1. Ensure the URL input is present and correctly formatted
    if (!urlInput) {
        return res.status(404).send('Not Found');
    }

    try {
        // 2. Parse the URL using a safe library function
        // The URL constructor parses the URL string
        const parsedUrl = new URL(urlInput);
        
        // 3. The network location attribute (hostname in Node URL API) is used to get the domain name
        // Redirect only if the domain strictly equals example.com
        if (parsedUrl.hostname === 'example.com') {
            return res.redirect(urlInput);
        } else {
            return res.status(404).send('Not Found');
        }
    } catch (err) {
        // If parsing fails, return a 404 error
        return res.status(404).send('Not Found');
    }
}

// Example usage context (e.g., Express route)
/*
app.get('/redirect', (req, res) => {
    redirect_url(req.query.url, res);
});
*/
module.exports = { redirect_url };
